### PR TITLE
Redefine countries regions

### PR DESCRIPTION
**Pull request by pabloarosado**
This PR creates a new dataset for countries-regions.

## Theoretical goals
1. We need to ensure that, across different datasets, regions mean the same thing. For example, if in one dataset China includes `Hong Kong` and `Macao` while another dataset doesn't, then they should not be called the same thing.
2. When creating aggregates (e.g. Europe) we need to ensure that we are adding data from all relevant regions and that we are not double-counting any region.

## Practical goal
Finding a solution for the previous goals may require a complicated data architecture and many hours of work. Or it may be impossible (given that data providers are usually unclear about region definitions). So we accept that there will always be some small geographical inaccuracies that are impossible to fix given our limitations. Given this, the goals become:
1. We need to ensure that inaccuracies are not very significant (e.g. we should not double-count Russia's contribution when aggregating data for Europe).
2. We need to save time when harmonizing country names and creating region aggregates.

## For reviewers
For major comments (e.g. "I think we should remove this table") it would be better to keep the conversation on [this issue](https://github.com/owid/etl/issues/779). For small comments, please use this PR.
* [x] @spoonerf regarding [this issue](https://github.com/owid/owid-issues/issues/13) currently the only aggregates we include are continents, EU 27 and the World (plus some other small geographical regions that happened to be in the old countries-regions dataset). Do you think there's any other region aggregates to include in this dataset (e.g. land-locked countries)? Otherwise, if those regions are created by specific institutions (e.g. income groups are defined by the World Bank and change periodically) I'd say they should be in different datasets.
* [x] @lucasrodes do you mind reviewing the `regions.py` garden step, `regions.meta.yml` metadata, and the `migration.py` script? The latter needs to be run only once, so it doesn't require a thorough review.
* [x] The `regions.codes.csv` doesn't need to be reviewed. It contains the legacy codes that were in the old dataset.
* [x] @bastianherre do you mind reviewing the content of `regions.yml`? Or if it's easier, you can review the definitions at the beginning of the `migrations.py` script. There you have some explanations (and you have country names instead of codes). If you need any clarifications, just let me know.
* [x] @Marigold I'd like to know if you agree with my comments below.

## Changes with respect to the old `countries_regions`
* New columns: `region_type`, `short_name`, `is_historical`, `successor`, `end_year`, `related`, and `defined_by`.
* New structure: Now we have one dataset, called `regions`, which contains the following tables:
  * `definitions`: Region definitions (name, country type, etc.).
  * `aliases`: Region aliases (i.e. variants of the region name).
  * `members`: Region members (roughly, sub-regions that would need to be added up when aggregating data for the region).
  * `related`: Other possible region members to be aware of (e.g. regions with an unclear official status that may be members of the region according to some data providers).
  * `transitions`: Historical transitions between regions.
  * `legacy_codes`: Legacy codes (e.g. wikidata code).
* The name of TLS has been corrected to "East Timor" (instead of "Timor").
* German and Italian subregions that were in the old dataset have been removed (I don't think we need country sub-regions for now). All other regions should be the same as in the old dataset.
* Some regions have been removed from the list of members of continents.
* Added a short name for the Democratic Republic of Congo (DR Congo).

## Minor comments
* I wasn't sure if the namespace should be `demography` or `geography`, so I simply called it `regions`, but let me know if anyone disagrees.
* I added the version `2023-01-01` just for simplicity, since it's unlikely that we will change it often.
* The idea here is that, if we want to include aliases or new regions, we can manually edit the `regions.yml` file directly, without creating a new dataset version. We should create a new dataset version only after a major change.
* We don't have to start replacing the `reference` dataset by this one for now. But we can start using this one when creating new datasets (after we agree on its content).
* This dataset will start making more sense once we create new helper functions for harmonization and to create region aggregates. For now, maybe we should not use this dataset.
* Currently the dataset only has one garden step that loads two adjacent files. This is an easy solution, but if we think it clashes too much with the ETL philosophy, I can move the two input files (`regions.yml` and `regions.codes.csv`) to snaphot, and the manual data gets ingested there.

### CODE REVIEW

File: etl/steps/data/garden/regions/2023-01-01/regions.yml
Comment (Marigold) [inline etl/steps/data/garden/regions/2023-01-01/regions.yml:6]: (We could have JSON schema validation to make sure we're not making mistakes when manually editing the file, but that could be added in the future)

**Review by Marigold — APPROVED** [review]
Can't really imagine cleaner way how to define regions. I hope it's gonna work well once we use it for harmonisation.

**Comment by Marigold** [comment]
Agree with everything, especially that we don't need snapshot for this and that we should be modifying existing version on minor updates.

> We don't have to start replacing the reference dataset by this one for now. But we can start using this one when creating new datasets (after we agree on its content).

It could be useful to create another PR that would use this dataset instead of reference (and compare its output with datadiff). Just to make sure that it works in practice and we didn't miss anything. (That doesn't block this PR though, it could be done in parallel or even after we merge this)

After we start using this file we'll have to modify [harmonize](https://github.com/owid/etl/blob/master/etl/harmonize.py#L265) to start saving aliases in the regions YAML file (`ruamel.yaml` is useful for modifying YAML files in-place). We can do it together with new helpers.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:571]: nice idea

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:734]: I would place this function before `_create_yaml_content_from_df`

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:776]: Duplicate row (see row 768-769)

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:780]: Duplicate rows (see rows 771-773)

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:802]: We could maybe assert that `sum(region_se) == 1`.

File: etl/steps/data/garden/regions/2023-01-01/regions.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/regions.py:32]: would place it after method `run`

File: etl/steps/data/garden/regions/2023-01-01/regions.py
Comment (lucasrodes) [inline etl/steps/data/garden/regions/2023-01-01/regions.py:50]: would place it after method `parse_raw_definitions`, after `run`.

**Review by lucasrodes — COMMENTED** [review]
LGTM! Thanks for working on this. Added minor comments.

**Review by lucasrodes — APPROVED** [review]

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (bastianherre) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:56]: Does this refer to Ethiopia before Eritrea split from it? I thought it was more commonly known as simply 'Ethiopia'.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (bastianherre) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:158]: Is this the official name for the country before it split into North and South? I thought it was more commonly known as just 'Korea'.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (bastianherre) [inline etl/steps/data/garden/regions/2023-01-01/migration.py:424]: Some datasets have data on these historical regions/states, before they became Germany and Italy — V-Dem is one prominent example. Would that be a reason to keep them?

**Review by bastianherre — COMMENTED** [review]
This is very nice work!

I left a few comments. Which sources did you use for all this information? Is this all based on UN data/information — or did you use other sources as well?

Please also let me know if I went about this review the wrong way — it is my first one!

**Comment by spoonerf** [comment]
Hey @pabloarosado - awesome that you have tackled this! 

Would it be worth adding in regions we are confident are unlikely to change in their composition? I can see that we might want to avoid the World Bank income groups here as these change periodically.

However, I think it could be valuable to add the World Bank regions and other organisational regions (WHO, UN) defined on [this page](https://ourworldindata.org/world-region-map-definitions) as I think these are unlikely to change. 

I think it is possible that the other groupings listed on [this issue](https://github.com/owid/owid-issues/issues/13) might change and therefore we wouldn't want to include them here.

**Review by spoonerf — APPROVED** [review]

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (danyx23) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:424]: Probably not - the two main uses of this file for us are a) to compute aggregations in a more automated way and b) to join data across datasets (e.g. when doing a scatterplot). I doubt that we have enough datasets that contain comparable definitions for these entities that either case is compelling. We can always _use_ countries within one dataset as we like - removing them here just communicates that we don't view these as useful enough that for either aggregation or joining. Does that make sense?

File: etl/steps/data/garden/regions/2023-01-01/regions.py
Comment (danyx23) [reply etl/steps/data/garden/regions/2023-01-01/regions.py:10]: I'd keep them as it might be useful in the future for updating old datasets or redoing datasets that come from literature. Wikidata codes might be useful at some point to fetch data from there (e.g. capitals for countries)

**Comment by danyx23** [comment]
Great work, @pabloarosado! This all looks very good. I have two points to add:

* I think we should consider if we really want the aliases to be stored in this file. The reason for this about ETL re-computation. In my mind, by far the most frequent change in the regions.yml file will be alias additions/changes. The way the ETL is currently set up this would then invalidated all downstream datasets (which would eventually be all, right?). There is a good chance that the added convenience and improved resilience is worth it but it's something I'd like to chat about in a data architecture meeting or similar.
* I think we should add a README.md file to this folder that explains how to make changes to the files (i.e. what is the threshold at which new entities should be added here, what are the fields we track and how to use them).

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py]: That's what the `migration.py` script does already. Or do you mean something else?

File: etl/steps/data/garden/regions/2023-01-01/regions.yml
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/regions.yml:34]: Good idea, I'm changing it now.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py]: Just reading from wikipedia and other sources, nothing extremely consistent, it was hard to find a consistent criterion...

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:812]: I just added a similar check above, ensuring that there are no repeated names.

File: etl/steps/data/garden/regions/2023-01-01/regions.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/regions.py:50]: I usually define `run` at the very end, and all additional functions are defined before `run`, and tend to appear in order of appearance in `run`. What's your criterion?

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:56]: Yes, I believe that's what it refers to. I didn't create this entity, but I suppose it was called `Eritrea and Ethiopia` to avoid confusion with `Ethiopia` current and former. But we could redefine `Eritrea and Ethiopia` -> `Ethiopia (former)`. Let me know if you think that's a better choice.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:158]: Yes, again, I don't know who introduced this definition in our dataset, but `United Korea` refers to the country before the split. I suppose it was called `United Korea` to avoid confusion. Feel free to propose a better name, e.g. `Korea (former)`.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:424]: Thanks Daniel, that's also how I see it. In fact I have been very permissive letting regions stay in the dataset, maybe I should have removed other complicated regions. But for this specific case, if we are not expecting these historical regions to appear in any dataset apart from V-Dem, then there's no strong reason to keep them.
In any case, if we change our mind, we can add them in the future.

**Comment by pabloarosado** [comment]
> This is very nice work!
> 
> I left a few comments. Which sources did you use for all this information? Is this all based on UN data/information, because some of this is controversial — or did you use other sources as well?
> 
> Please also let me know if I went about this review the wrong way — it is my first one!

Thank you @bastianherre! Regarding sources, I couldn't find one specific source where all the required definitions were clearly stated. So I used various sources, mostly Wikipedia. If you have found something that is particularly controversial, or that we should change, please let me know.

And yes, you did well with the review, adding comments by line is very useful, thanks!

**Comment by pabloarosado** [comment]
> Hey @pabloarosado - awesome that you have tackled this!
> 
> Would it be worth adding in regions we are confident are unlikely to change in their composition? I can see that we might want to avoid the World Bank income groups here as these change periodically.
> 
> However, I think it could be valuable to add the World Bank regions and other organisational regions (WHO, UN) defined on [this page](https://ourworldindata.org/world-region-map-definitions) as I think these are unlikely to change.
> 
> I think it is possible that the other groupings listed on [this issue](https://github.com/owid/owid-issues/issues/13) might change and therefore we wouldn't want to include them here.

Hi @spoonerf thanks for the clarification and the suggestion. To avoid blocking this one, I've created [a separate issue](https://github.com/owid/owid-issues/issues/945) to add those regions.

**Comment by pabloarosado** [comment]
> Great work, @pabloarosado! This all looks very good. I have two points to add:
> 
> * I think we should consider if we really want the aliases to be stored in this file. The reason for this about ETL re-computation. In my mind, by far the most frequent change in the regions.yml file will be alias additions/changes. The way the ETL is currently set up this would then invalidated all downstream datasets (which would eventually be all, right?). There is a good chance that the added convenience and improved resilience is worth it but it's something I'd like to chat about in a data architecture meeting or similar.
> * I think we should add a README.md file to this folder that explains how to make changes to the files (i.e. what is the threshold at which new entities should be added here, what are the fields we track and how to use them).

Hi @danyx23 thanks!
* Regarding aliases, I'm not sure what the best solution is. If you move the aliases outside of this dataset to another dataset, then that new dataset would have the same problem, wouldn't it? We could simply omit that dependency, but that would not be an ideal solution. I think aliases don't change that often anymore, so I don't think it's a big deal. Locally, we usually have to run only the steps that we need to work on, and in production, all datasets are executed regularly anyway. What do you think? We can talk about it tomorrow during the data architecture chat.
* Regarding the README, no problem. I can add one to `etl/docs/data`. It's quite a hidden place (and there's only one more dataset README there) but I think it's probably the right place. Let me know if you have a better idea.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (bastianherre) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:424]: Yes, that makes sense. I expect us to add more datasets with these historical states, and we may want to create some aggregates there — but the number of datasets probably will likely stay low.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (bastianherre) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:56]: Yes, if we want a different name for when Eritrea was part of Ethiopia, I would prefer Ethiopia (former). That would make clear that it was indeed one country.

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (pabloarosado) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:424]: I don't think it's a problem to include those regions, but we'd simply need to add the necessary information (e.g. their end year). But we can do this as a separate issue, to avoid delaying this one. It would be better to add those regions once we have better tools in place (i.e. `harmonize` and `add_region_aggregates` are updated).

File: etl/steps/data/garden/regions/2023-01-01/migration.py
Comment (bastianherre) [reply etl/steps/data/garden/regions/2023-01-01/migration.py:158]: In this case, I would not use '(former)', as some may understand that as there being a 'Korea' in the present. I would simply use 'Korea', which is the name usually used.

**Comment by pabloarosado** [comment]
Hi @bastianherre, so you are suggesting two changes:
* `Eritrea and Ethiopia` -> `Ethiopia (former)`.
* `United Korea` -> `Korea`.

We have [another issue](https://github.com/owid/etl/issues/156) where we proposed to use `(former)` for all historical regions. But lately I've been thinking that it may not be a good idea. For example, `USSR (former)` may also sound like there is another `USSR` now.

I created [a separate issue](https://github.com/owid/etl/issues/796) for renaming countries, and people were generally quite reticent to do so. But these two changes you suggested are used probably in very few datasets, so it shouldn't be a big problem. If you want, feel free to propose the changes as a new comment in that issue, or in Slack, and (if others agree) I can implement those changes in a separate pull request (for the moment we are not using the `regions` dataset anywhere, so it should be fine to update it again soon). Thanks.

**Comment by pabloarosado** [comment]
Final changes:
* I've implemented all suggestions above.
* I've realised we still had `Caribbean Netherlands`, which, as we mentioned in the past, should not be included as a separate country, since it's the same region as the BES Islands. Therefore, I added `Caribbean Netherlands` in the list of aliases of `Bonaire Sint Eustatius and Saba`.
* I added a README in `docs/data/regions.md`, @danyx23 feel free to take a look.

**Pull request merged**
